goose.go

```diff
@@ -1051,136 +1065,142 @@ func (c *cursor) Remainder() []ast.Stmt {
+// ifStmt has special support for an early-return "then" branch; to achieve that
+// it is responsible for generating the if *together with* all the statements that
+// follow it in the same block.
+// It is also responsible for "finalizing" the expression according to the usage.
+func (ctx Ctx) ifStmt(s *ast.IfStmt, remainder []ast.Stmt, usage ExprValUsage) coq.Binding {
 	// TODO: be more careful about nested if statements; if the then branch has
 	//  an if statement with early return, this is probably not handled correctly.
 	//  We should conservatively disallow such returns until they're properly analyzed.
 	if s.Init != nil {
 		ctx.unsupported(s.Init, "if statement initializations")
 		return coq.Binding{}
 	}
-	thenExpr, ok := ctx.stmt(s.Body, &cursor{nil}, loopVar).Unwrap()
-	if !ok {
-		ctx.nope(s.Body, "if statement body ends with an assignment")
-		return coq.Binding{}
-	}
+	condExpr := ctx.expr(s.Cond)
 	ife := coq.IfExpr{
-		Cond: ctx.expr(s.Cond),
-		Then: thenExpr,
-	}
-
-	// supported use cases
-	// (1) early return, no else branch; remainder of function is lifted to else
-	// (2) both branches and no remainder
-	//
-	// If the else branch also doesn't return it's not a problem to handle subsequent code,
-	// but that's annoying and we can leave it for later. Maybe the special case
-	// of Else == nil should be supported, though.
-
-	remaining := c.HasNext()
-	bodyEndsWithReturn := endsWithReturn(s.Body)
-	if bodyEndsWithReturn && remaining {
-		if s.Else != nil {
-			ctx.futureWork(s.Else, "else with early return")
-			return coq.Binding{}
-		}
-		ife.Else = ctx.stmts(c.Remainder(), loopVar)
-		return coq.NewAnon(ife)
-	}
-	if !bodyEndsWithReturn && remaining && s.Else == nil {
-		// conditional statement in the middle of a block
-		retUnit := coq.ReturnExpr{coq.Tt}
-		ife.Then = coq.BlockExpr{[]coq.Binding{
-			coq.NewAnon(ife.Then),
-			coq.NewAnon(retUnit),
-		}}
-		ife.Else = retUnit
-		return coq.NewAnon(ife)
-	}
-	elseEndsWithReturn := endsWithReturn(s.Else)
-	// the if expression is last in the surrounding block,
-	// so returns in it become returns from the overall function
-	// OR
-	// neither branch terminates the outer function,
-	// the if expression is just a conditional in the middle
-	if !remaining || (!bodyEndsWithReturn && !elseEndsWithReturn) {
-		if s.Else == nil {
-			ife.Else = coq.ReturnExpr{coq.Tt}
-			return coq.NewAnon(ife)
-		}
-		elseExpr, ok := ctx.stmt(s.Else, c, loopVar).Unwrap()
-		if !ok {
-			ctx.nope(s.Else, "if statement else ends with an assignment")
-			return coq.Binding{}
+		Cond: condExpr,
+	}
+
+	// Determine if we want to propagate our usage (i.e. the availability of control effects)
+	// to the conditional or not. This is not relevant for correctness (stmt generation will
+	// appropriately error when a control effect arises where it is not supported), but Crucial
+	// for supporting a reasonable subset of Go.
+	bodyEndsWithReturn := ctx.endsWithReturn(s.Body)
```

Comment:
This is a crucial part of this approach -- we still use `endsWithReturn` to decide how to translate conditionals, but the worst thing that can happen when it gives a wrong answer is that we end up with a control effect where none is supported, or in the "early return in if with an else branch" below. But the behavior of `endsWithReturn` should be entirely irrelevant for correctness now.

Reply:
Actually I just realized this is not correct; if `endsWithReturn` returns `true`, then it *must* always return. Otherwise putting the remainder into the "else" branch is wrong.

It would be okay for `endsWithReturn` to spuriously return `false`, but that would definitely lead to a translation error since we'd use `ExprValLocal` translation.